Add endpoint serving static maps (by stitching tiles)

petrsloup opened the pull request:
This PR adds `/static/{lon},{lat},{zoom}/{width}x{height}[@{scale}x].{format}` endpoint (syntax similar to https://www.mapbox.com/developers/api/static/) which serves requested static image.

The static image is stitched from individual tiles using https://github.com/mapbox/abaculus.

I had to do some refactoring (extracting code from the tile route into `getTile` function) to prevent code duplicity and slightly modify related error reporting. Hope this was the right way to handle this and it doesn't break anything.

klokan commented:
@mojodna would you have time for a short review of this PR, Seth?
I believe it is a clean and nice code adding an interesting feature...

mojodna commented on lib/app.js:
```diff
@@ -11,6 +11,8 @@ var cachecache = require("cachecache"),
     handlebars = require("handlebars"),
     mercator = new (require("sphericalmercator"))();
 
+var abaculus = require('abaculus');
```
Please move this into the alphabetized list of "community" dependencies (the block above).

mojodna commented on lib/app.js:
```diff
@@ -122,48 +124,41 @@ module.exports = function(tilelive, options) {
-          return res.status(404).end();
+        if (ext !== format) {
+          debug("Invalid format '%s', expected '%s'", format, ext);
+          return callback(404, 'Invalid format', populateHeaders({}, params, { 404: true, invalidFormat: true }));
```
This (and the subsequent callbacks) should pass an `Error` instance as the first argument for consistency.

mojodna commented on lib/app.js:
```diff
@@ -205,11 +197,55 @@ module.exports = function(tilelive, options) {
+  app.get(tilePattern, function(req, res, next) {
+    return getTile(req.params.z | 0, req.params.x | 0, req.params.y | 0,
+        (req.params.scale || "@1x").slice(1, 2), req.params.format,
+        function(status, data, headers) {
```
Can you expand the inline variable manipulation into separate variables above this? (For readability)

mojodna commented on lib/app.js:
```diff
@@ -205,11 +197,55 @@ module.exports = function(tilelive, options) {
+    return getTile(req.params.z | 0, req.params.x | 0, req.params.y | 0,
+        (req.params.scale || "@1x").slice(1, 2), req.params.format,
+        function(status, data, headers) {
+          if (status && Number(status) == NaN) {
```
Indentation in this block is wrong.

mojodna replied:
As above, please switch to `Error` instances for consistency (with a `status` property). The first argument to a callback should always be an optional error (`null` or `undefined` if no error occurred).

This then becomes something like:

```javascript
if (err && !err.status) {
  return next(err);
}

// ...
```

In writing this, I'm thinking that 404s should be treated as non-errors and detected by checking whether `data == null` (`data === null || data === undefined`), similar to the straightforward 404 check above.

mojodna commented on lib/app.js:
```diff
@@ -205,11 +197,55 @@ module.exports = function(tilelive, options) {
+    ":width(\\d+)x:height(\\d+):scale(@[123]x)?\.:format([\\w\\.]+)";
+
+  app.get(staticPattern, function(req, res, next) {
+    var scale = (req.params.scale || "@1x").slice(1, 2) | 0;
```
Please merge the `var` definitions to match style elsewhere.

Also consider extracting the `scale` determination as `getScale(scale)`, since this appears more than once now (and isn't totally straightforward).

mojodna commented on lib/app.js:
```diff
@@ -205,11 +197,55 @@ module.exports = function(tilelive, options) {
+      zoom: req.params.z | 0,
+      scale: scale,
+      center: {
+        x: parseFloat(req.params.lon),
```
Local style for float coercion is `+req.params.lon` (similar to `req.params.width | 0` being the local idiom used for int coercion).

mojodna commented:
This is awesome; it's something I've been meaning to do myself for a while. Sorry for my delay in reviewing it.

My primary concern is the substitution of numbers for `Error` objects, since that deviates from Node callback conventions. The remaining stuff isn't crucial, but I'd appreciate if you could clean it up (otherwise I will and it may delay things more).

Would you consider adding bounding box static maps, per http://docs.cartodb.com/cartodb-platform/maps-api/static-maps-api/#bounding-box ?  IIRC, Abaculus supports bounding boxes in place of center/zoom.

mojodna commented on lib/app.js:
```diff
@@ -205,11 +197,55 @@ module.exports = function(tilelive, options) {
+  var floatPattern = "[+-]?(?:\\d+|\\d+\.?\\d+)";
+  var staticPattern =
+    "/static/:lon(" + floatPattern + "),:lat(" + floatPattern + "),:z(\\d+)/" +
+    ":width(\\d+)x:height(\\d+):scale(@[123]x)?\.:format([\\w\\.]+)";
```
I tend to prefer `util.format` for string manipulation, though I could go either way here.

Consider lifting `floatPattern` into a top-level "constant" in case it could be used elsewhere.

The scale pattern (`@[123]x`) could be similarly extracted as `SCALE_PATTERN`, since it's also used in the tile endpoint.

petrsloup commented:
@mojodna Thanks for the review and remarks ! I believe I addressed all the codestyle-related issues.

I also added the bbox-based endpoint, but I'm not sure about the URL syntax.
When we have full bbox, it does not make sense to specify both `width` and `height` + abaculus is probably not ready for scaling images.

So, this is the implemented syntax at the moment:
`/static/{minx},{miny},{maxx},{maxy}/{zoom}[@{scale}x].{format}`

mojodna commented on lib/app.js:
```diff
@@ -15,6 +17,9 @@ var tessera = require("./index");
 debug = debug("tessera");
 
+var FLOAT_PATTERN = "[+-]?(?:\\d+|\\d+\.?\\d+)";
+var SCALE_PATTERN = "@[123]x";
```
Minor nitpick: `1` shouldn't be in this list.

mojodna commented on lib/app.js:
```diff
@@ -205,11 +204,85 @@ module.exports = function(tilelive, options) {
+          format: format,
+          getTile: function(z, x, y, callback) {
+            return getTile(z, x, y, scale, format, function(err, data, headers) {
+              if (!err && data == null) {
```
You probably want to keep `data` as null, otherwise abaculus will fail to stitch maps that are partially transparent (due to tiles that aren't present):

https://github.com/mapbox/abaculus/blob/master/index.js#L177
https://github.com/mapbox/abaculus/blob/master/index.js#L159

petrsloup replied:
You're right, but I can't think of any easy way to achieve this. Passing `null` data to abaculus would cause it to fail; not calling the callback would cause waiting until timeout. Only valid option is to either patch this somehow in abaculus or create empty 1x1 transparent image and pass this instead of the 404, but we would probably need to have these in various formats.

The only drawback of this behavior is that we return 404 even if only some of the required tiles are missing. I think this does not block the basic usage and does not need to be fixed in this PR.

mojodna commented on lib/app.js:
```diff
@@ -205,11 +204,85 @@ module.exports = function(tilelive, options) {
+        y = req.params.y | 0,
+        scale = getScale(req.params.scale),
+        format = req.params.format;
+    return getTile(z, x, y, scale, format, function(err, data, headers) {
```
This lost the behavior whereby `data == null` tiles are treated as 404s (due to the nice cleanup above).

petrsloup replied:
Yes, thanks. Fixed with https://github.com/klokantech/tileserver-tessera/commit/fb3ddfa6d259ef7f81f52e718f613dab6b994663

mojodna commented:
@petrsloup you rock. This is almost ready to go.

Abaculus does appear to scale images: https://github.com/mapbox/abaculus/blob/master/index.js#L187-L188

So perhaps width/height should be optional? (Though not necessary now (plus, it does ugly things to labels), we can do that later if/when one of us needs it.)

petrsloup commented:
@mojodna When passing `bbox` to abaculus, it also needs the zoom level to be specified. If we want `width`x`height` in urls, we would need to calculate the required zoom level (of the data to be loaded) + ensure the scaling.

I think this is out of the scope of this PR and can be done later.

mojodna commented:
:+1: on both additions being out of scope.

Thanks! I'll get a release out with this ASAP.

petrsloup commented:
Great, thanks !

mojodna commented:
Published as v0.6.0. Sorry for the delay.
